### PR TITLE
feat(relayforreddit): add `change-client-id` patch

### LINKED ISSUE #753: feat: Relay for reddit patch
### Application

Relay Pro for reddit 

### Issue

I prefer it over sync, which isn't exactly a good reason. A good reason is the developer Dbrady isn't abandoning the app, but is going to try for a subscription model.

### Patch

Oauth patch

### Motivation

Continue to use the app that I've used for a decade, and will have original support from the developer. 

### Acknowledgements

- [X] I have searched the existing issues and this is a new and no duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this issue properly.

### PR DESCRIPTION
**Pull request by gudmunduro**
Add a patch to change client ID in `Relay for reddit`

closes: ReVanced/revanced-patches-template#754

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.reddit.customclients.relayforreddit.api.fingerprints.GetClientIdFingerprint
+
+@Patch
+@Name("change-client-id")
+@Description("Changes the OAuth client ID")
+@Compatibility([Package("free.reddit.news"), Package("reddit.news")])
+@Version("0.0.1")
+class ChangeClientIdPatch :
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt:22]: Please format the class.

File: src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import app.revanced.patcher.patch.OptionsContainer
+import app.revanced.patcher.patch.PatchOption
+import app.revanced.patcher.patch.PatchResultError
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.reddit.customclients.relayforreddit.api.fingerprints.GetClientIdFingerprint
+
+@Patch
+@Name("change-client-id")
+@Description("Changes the OAuth client ID")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt:19]:
```suggestion
@Description("Changes the OAuth client ID.")
```

File: src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+@Compatibility([Package("free.reddit.news"), Package("reddit.news")])
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt:20]: It has two?

File: src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt (hunk 1 above)
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt:22]: Extend from `AbstractClientIdPatch`, follow the same style as the existing Reddit client patches.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt (hunk 3 above)
Comment (gudmunduro) [reply src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt:22]: Separate app for free and pro version

File: src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import app.revanced.patcher.patch.PatchOption
+import app.revanced.patcher.patch.PatchResultError
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.reddit.customclients.AbstractChangeOAuthClientIdPatch
+import app.revanced.patches.reddit.customclients.relayforreddit.api.fingerprints.GetClientIdFingerprint
+
+@Patch
+@Name("change-oauth-client-id")
+@Description("Changes the OAuth client ID in Relay for Reddit.")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt:21]: The user already knows which app they patch
```suggestion
@Description("Changes the OAuth client ID.")
```

File: src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+@Name("change-oauth-client-id")
+@Description("Changes the OAuth client ID.")
+@Compatibility([Package("free.reddit.news"), Package("reddit.news")])
+@Version("0.0.1")
+class ChangeOAuthClientIdPatch : AbstractChangeOAuthClientIdPatch(
+    "dbrady://relay",
+    Options,
+    GetClientIdFingerprint
+) {
+
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt:29]:
```suggestion
```

File: src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    "dbrady://relay",
+    Options,
+    GetClientIdFingerprint
+) {
+    override fun MethodFingerprint.patch(context: BytecodeContext): PatchResult {
+        val result = GetClientIdFingerprint.result ?: return GetClientIdFingerprint.toErrorResult()
+
+        result.mutableMethod.replaceInstruction(0, "const-string v0, \"$clientId\"")
+
+        return PatchResultSuccess()
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/relayforreddit/api/patch/ChangeClientIdPatch.kt:34]:
```suggestion
        val result = GetClientIdFingerprint.result?.mutableMethod?.replaceInstruction(0, "const-string v0, \"$clientId\"")
            ?: return GetClientIdFingerprint.toErrorResult()

        return PatchResultSuccess()
```

**Review by oSumAtrIX — APPROVED** [review]

**Review by oSumAtrIX — APPROVED** [review]

**Comment by Toiletmason** [comment]
@gudmunduro Nice work my man. I would open this up to the pro version of the app. The free version is going to be dropped, while the pro version will continue to receive updates as far as I have heard.

**Comment by oSumAtrIX** [comment]
Is this a runtime error and what's the stack trace?

**Comment by gudmunduro** [comment]
> @gudmunduro Nice work my man. I would open this up to the pro version of the app. The free version is going to be dropped, while the pro version will continue to receive updates as far as I have heard.

This should also work for the pro version

**Comment by toadboi98** [comment]
> pro

I am having an issue patching the pro version (reddit.news.apk) which I downloaded from the playstore. Any suggestions? 

`SEVERE: Change OAuth client id failed: app.revanced.patcher.patch.PatchResultError: Failed to resolve LoginActivityClientIdFingerprint`

**Comment by oSumAtrIX** [comment]
Open an issue.

**Comment by brian6932** [comment]
Sorry for the bump, but were you ever able to patch the Pro version (`reddit.news` not `free.reddit.news`)? @toadboi98
I can patch the normal APK just fine, but for some reason the Pro patch doesn't work for me, it patches without any errors, but doesn't seem to use the CID, and just says `Failed to load content` (both APKs are on version `10.2.40`)

Due to the lack of response, I'll open an issue regarding this

**Pull request merged**